### PR TITLE
Add BM25STD scorer + ScoringSession for full-text scoring

### PR DESCRIPTION
**Pull request by boda26**
## Summary

  Adds the math + per-query collection layer for BM25STD scoring.
  Algorithm-agnostic `Scorer` / `ScoringStats` interfaces so TFIDF
  (and later tag / numeric scorers) can plug in without changes.
  Driving the session from the predicate tree is a follow-up.

  ## Changes

  - `src/indexes/scoring/scoring_stats.h` — `ScoringStats` base +
    `Bm25StdStats` / `TfidfStats` subclasses
  - `src/indexes/scoring/scorer.{h,cc}` — `Scorer` interface,
    default `CombineGroup`
  - `src/indexes/scoring/bm25std_scorer.{h,cc}` — Okapi BM25
    (k1=1.2, b=0.75). Uses a bit-pattern `IsInf` because
    `-ffast-math` makes `std::isinf` unreliable.
  - `src/indexes/scoring/scoring_session.{h,cc}` — per-query
    collector. `RecordLeaf` / `EnterGroup` / `ExitGroup` / `Rank`.
  - Wired as `scoring` static lib in `src/indexes/CMakeLists.txt`.

  ## Tests

  New `scoring_test` binary (26 tests, all pass):

  - `testing/scoring/scoring_test_data.h` — shared 8-doc, 4-term
    corpus designed to isolate TF saturation, length norm, IDF
    differentiation, tie-break, partial-match admission.
  - `testing/scoring/bm25std_scorer_test.cc` — scorer math,
    edge cases, death tests for invariant violations.
  - `testing/scoring/scoring_session_test.cc` — all query shapes
    the corpus unlocks (single leaf, AND, OR, weighted/nested
    groups, mixed admission, ties). Absolute scores verified
    against Redis 8.6 docker.

### CODE REVIEW

File: src/indexes/scoring/scoring_stats.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+//   positions  - positional offsets of `term` within the document, sorted
+//                ascending; empty when SLOP is not used
+struct TfidfStats : ScoringStats {
+  std::unique_ptr<ScoringStats> Clone() const override {
+    return std::make_unique<TfidfStats>(*this);
+  }
+
+  uint32_t norm = 0;
+  absl::Span<const uint32_t> positions;
+};
```
[the remark this answers is not in the document]
Comment (boda26) [reply src/indexes/scoring/scoring_stats.h:85]: This PR focuses on BM25STD implementation. Will fix this later in the TFIDF implementation.

File: src/indexes/scoring/scoring_stats.h [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+//   term_frequency        - doc-wide term frequency
+struct ScoringStats {
+  virtual ~ScoringStats() = default;
+
+  // Polymorphic deep copy. ScoringSession uses this to take an owned
+  // copy of caller-supplied stats so its lifetime is independent of
+  // the original.
+  virtual std::unique_ptr<ScoringStats> Clone() const {
+    return std::make_unique<ScoringStats>(*this);
+  }
```
[the remark this answers is not in the document]
Comment (boda26) [reply src/indexes/scoring/scoring_stats.h:40]: This PR focuses on BM25STD implementation. Will fix this later in the TFIDF implementation.

**Comment by boda26** [comment]
Need a major change for all variables from double to float. Closing this PR and will open a new one later
